### ISSUE TITLE
how should posit flexibility and simplicity be balanced?

### ISSUE THREAD

**Issue by agshew** [posit-standard/posit-standard#5]
John Gustafson wrote:
> I would appreciate suggestions about how to resolve the dilemma of wanting a simple standard but also wanting the full flexibility of various es values for any nbits precision. As you see, I elected to restrict the es values, which greatly simplifies the formulas. Do we need a separate Standard for a fully flexible posit? When it comes to language support, I think it's going to be a nightmare to have more than a dozen kinds of posits, each requiring its own math library support. But I'm up for it if the hardware people tell me it won't put posits at a disadvantage to IEEE floats for performance.

**Comment by xman** [comment]
Fix the es values for currently supported nbits as in the current draft.
Other potential nbits may be considered in the future as we have more use case and data to support.

Suggest not to include formulae to derive the es values, so that we have the flexibility to decide the es values of different nbits in the future which may break the current es formula.

**Comment by Ravenwater** [comment]
I agree with @xman on the desire to explicitly use nbits and es to specify the posit configuration.

The small posit configurations, with nbits in the range of 2 up to 8, are super interesting for applications that leverage the quantization of these posits: AI being the poster child. I am exploring uses in control and IoT.

If you look at use cases where you start with a required precision or sample coverage, but you need to experiment with dynamic range, then the progression for these small posit configuration would move the es value up. For example, <7,0>, <8,1>, <9,2>. For data path generators this is attractive as you keep the critical path of the arithmetic part constant, and thus you can get an idea of performance of any of these progressions by studying the <7,0> config (cycle time, gate count, resource utilization, power consumption, etc.)

As there is a trend away from general purpose computing towards specialized hardware acceleration, there is an opportunity for driving new levels of storage and compute efficiency by using custom tailored posit configurations.  The standard posit configurations of 8/16/32/64 appear to coarse to support that trend. I would like to see 4/8/16/24/32/48/56/64/80/96/112/128. As most linear algebra is memory bound, having these 'smaller' posits available, say a <48,2>, <56,3> to reduce the memory and communication bandwidth would be a key differentiator for HPC applications.

The one difficulty I do run into is the sw complexity of the conversion requirement of the standard. Just providing conversions and comparisons between posits and the existing standard types is creating a lot of additional code. for a library or compiler The 'full' posit configuration state space is ginorous, and implicit conversions create subtle bugs that we all know from experience are impossible to find in large codes. New languages like Golang have explicitly removed implicit conversions.

This first language that is showcasing this problem is Python. It does not appear reasonable to expose general posit configurations into Python, and thus Python applications will be constrained to use the standard configurations.
